// dw/catalog/CatalogMgr.js
/**
 * Sample demo for teaching xml parsing and conversion using xml-js library in a Salesforce B2C Commerce context
 */
'use strict';

const categories = require('./categories.json');
const categoryIndex = Object.keys(categories);

module.exports.getSiteCatalog = function() {
    return {
        ID : 'miditech-storefront-catalog-fr',
        root : module.exports.getCategory('root')
    }
}

module.exports.getCategory = function(cid) {

    if(cid !== 'root' && categoryIndex.indexOf(cid) < 0) {
        return null;
    }

    return {
        ID : cid,
        products : DwCollection({
            size : categories[cid] || 0,
            buildNext : function(cursor) {
                return {
                    ID : 'product-'+cursor
                }
            }
        }),    
        subCategories  : DwCollection({
            size : cid === 'root' ? categoryIndex.length : 0,
            buildNext : function(cursor) {
                return module.exports.getCategory(categoryIndex[cursor]);
            }
        })
    }

}

function DwCollection(opts) {
    let cur = -1;
    const size = opts.size;
    return {
        size : function() { return size },
        iterator : function() {
            const _iterator = {
                hasNext : function() { return (cur+1) < size },
                next : function() {
                    if(!_iterator.hasNext()) {
                        throw new Error('java.util.NoSuchElementException : no more element to iterate');
                    }
                    cur++;
                    return opts.buildNext(cur);
                }
            }

            return _iterator;
        }
    }
}
